File: app/src/main/java/sh/siava/AOSPMods/utils/batteryStyles/CircleBatteryDrawable.java
package sh.siava.AOSPMods.utils.batteryStyles;

import static android.graphics.Color.WHITE;
import static android.graphics.Paint.ANTI_ALIAS_FLAG;
import static android.graphics.Paint.Align.CENTER;
import static android.graphics.Paint.Style.STROKE;
import static android.graphics.Typeface.BOLD;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.Resources;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.DashPathEffect;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.PathEffect;
import android.graphics.PixelFormat;
import android.graphics.RectF;
import android.graphics.SweepGradient;
import android.graphics.Typeface;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.graphics.ColorUtils;
import androidx.core.graphics.PathParser;

public class CircleBatteryDrawable extends BatteryDrawable
{
	private static final int BATTERY_STYLE_CIRCLE = 1;
	private static final int BATTERY_STYLE_DOTTED_CIRCLE = 2;
	private static final String WARNING_STRING = "!";
	private static final int CRITICAL_LEVEL = 5;
	private static final int CIRCLE_DIAMETER = 45; //relative to dash effect size. Size doesn't matter as finally it gets scaled by parent
	private static final PathEffect DASH_PATH_EFFECT = new DashPathEffect(new float[]{3f, 2f}, 0f);
	private final Context mContext;
	private boolean mIsFastCharging = false;
	private boolean mIsCharging = false;
	private boolean mIsPowerSaving = false;
	private boolean mShowPercentage = false;
	private int mBatteryLevel;
	private int mDiameter;
	private final RectF mFrame = new RectF();
	private int mFGColor = WHITE;
	private final Paint mTextPaint = new Paint(ANTI_ALIAS_FLAG);
	private final Paint mFramePaint = new Paint(ANTI_ALIAS_FLAG);
	private final Paint mBatteryPaint = new Paint(ANTI_ALIAS_FLAG);
	private final Paint mWarningTextPaint = new Paint(ANTI_ALIAS_FLAG);
	private final Paint mBoltPaint = new Paint(ANTI_ALIAS_FLAG);
	private final Paint mPowerSavePaint = new Paint(ANTI_ALIAS_FLAG);
	private int[] mShadeColors;
	private float[] mShadeLevels;
	private long mLastUpdate;
	private int mAlpha = 255;
	private Path mBoltPath;

	@SuppressLint("DiscouragedApi")
	public CircleBatteryDrawable(Context context, int frameColor)
	{
		mContext = context;
		Resources res = mContext.getResources();

		mFramePaint.setDither(true);
		mFramePaint.setStyle(STROKE);

		mTextPaint.setTypeface(Typeface.create("sans-serif-condensed", BOLD));
		mTextPaint.setTextAlign(CENTER);

		mWarningTextPaint.setTypeface(Typeface.create("sans-serif", BOLD));
		mWarningTextPaint.setTextAlign(CENTER);

		mBatteryPaint.setDither(true);
		mBatteryPaint.setStyle(STROKE);

		int mPowerSaveColor = getColorStateListDefaultColor(mContext, res.getIdentifier(
				"batterymeter_plus_color",
				"color",
				mContext.getPackageName()));

		mPowerSavePaint.setColor(mPowerSaveColor);
		mPowerSavePaint.setStyle(STROKE);

		setColors(frameColor, frameColor, frameColor);

		setMeterStyle(BATTERY_STYLE_CIRCLE);
	}

	@Override
	public void setShowPercent(boolean showPercent) {
		mShowPercentage = showPercent;
		postInvalidate();
	}

	@Override
	public void setMeterStyle(int batteryStyle) {
		mFramePaint.setPathEffect(batteryStyle == BATTERY_STYLE_DOTTED_CIRCLE ? DASH_PATH_EFFECT : null);
		mBatteryPaint.setPathEffect(batteryStyle == BATTERY_STYLE_DOTTED_CIRCLE ? DASH_PATH_EFFECT : null);
		mPowerSavePaint.setPathEffect(batteryStyle == BATTERY_STYLE_DOTTED_CIRCLE ? DASH_PATH_EFFECT : null);

		postInvalidate();
	}

	@Override
	public void setFastCharging(boolean isFastCharging) {
		mIsFastCharging = isFastCharging;
		if(isFastCharging) mIsCharging = true;

		postInvalidate();
	}

	private void postInvalidate()
	{
		unscheduleSelf(this::invalidateSelf);
		scheduleSelf(this::invalidateSelf, 0);
	}

	@Override
	public void setCharging(boolean isCharging) {
		mIsCharging = isCharging;
		if(!isCharging) mIsFastCharging = false;

		postInvalidate();
	}

	@Override
	public void setBatteryLevel(int level) {
		mBatteryLevel = level;
		postInvalidate();
	}

	@Override
	public void setBounds(int left, int top, int right, int bottom)
	{
		super.setBounds(left, top, right, bottom);
		updateSize();
	}

	@Override
	public void setColors(int fgColor, int bgColor, int singleToneColor) {
		mFGColor = fgColor;

		mBoltPaint.setColor(mFGColor);
		mFramePaint.setColor(bgColor);
		mTextPaint.setColor(mFGColor);

		setAlpha(mAlpha); //invalidate included here - alpha shall be re-applied after color change
	}

	@Override
	public void setPowerSaving(boolean isPowerSaving) {
		mIsPowerSaving = isPowerSaving;

		postInvalidate();
	}

	@Override
	public void refresh() {
		postInvalidate();
	}

	private void refreshShadeColors()
	{
		if(batteryColors == null) return;

		mShadeColors = new int[batteryLevels.size() * 2 + 2];
		mShadeLevels = new float[mShadeColors.length];

		float lastPCT = 0f;

		for(int i = 0; i < batteryLevels.size(); i++)
		{
			float rangeLength = batteryLevels.get(i) - lastPCT;

			int pointer = 2 * i;
			mShadeLevels[pointer] = (lastPCT + rangeLength * 0.3f) / 100;
			mShadeColors[pointer] = batteryColors[i];

			mShadeLevels[pointer + 1] = (batteryLevels.get(i) - rangeLength * 0.3f) / 100;
			mShadeColors[pointer + 1] = batteryColors[i];
			lastPCT = batteryLevels.get(i);
		}

		mShadeLevels[mShadeLevels.length - 2] = (batteryLevels.get(batteryLevels.size() - 1) + (100 - batteryLevels.get(batteryLevels.size() - 1) * 0.3f)) / 100;
		mShadeColors[mShadeColors.length -2] = Color.GREEN;

		mShadeLevels[mShadeLevels.length - 1] = 1f;
		mShadeColors[mShadeColors.length- 1] = Color.GREEN;
	}

	@Override
	public void draw(@NonNull Canvas canvas) {
		if(mBatteryLevel == -1 || mDiameter == 0) return;

		if (mLastUpdate != lastVarUpdate)
		{
			mLastUpdate = lastVarUpdate;
			refreshShadeColors();
		}

		if(!mIsPowerSaving)
			setLevelBasedColors(mBatteryPaint, mFrame.centerX(), mFrame.centerY());

		if(mIsCharging)
		{
			canvas.drawPath(mBoltPath, mBoltPaint);
		}

		canvas.drawArc(mFrame, 270f, 360f, false, mFramePaint);

		if(mBatteryLevel > 0)
		{
			canvas.drawArc(mFrame, 270f, 3.6f * mBatteryLevel, false,
					(!mIsCharging && mIsPowerSaving)
						? mPowerSavePaint
						: mBatteryPaint);
		}

		if(!mIsCharging && mBatteryLevel < 100 && mShowPercentage)
		{
			String pctText = mBatteryLevel > CRITICAL_LEVEL ? String.valueOf(mBatteryLevel) : WARNING_STRING;

			float textHeight = -mTextPaint.getFontMetrics().ascent;
			float pctX = mDiameter * .5f;
			float pctY = (mDiameter + textHeight) * 0.47f;
			canvas.drawText(pctText, pctX, pctY, mTextPaint);
		}
	}

	private void setLevelBasedColors(Paint paint, float centerX, float centerY)
	{
		paint.setShader(null);

		if(mIsFastCharging && showFastCharging && mBatteryLevel < 100)
		{
			paint.setColor(fastChargingColor);
			return;
		} else if (mIsCharging && showCharging && mBatteryLevel < 100) {
			paint.setColor(chargingColor);
			return;
		}

		if(!colorful || mShadeColors == null)
		{
			for(int i = 0; i < batteryLevels.size(); i++)
			{
				if(mBatteryLevel <= batteryLevels.get(i))
				{
					if(transitColors && i > 0)
					{
						float range = batteryLevels.get(i) - batteryLevels.get(i - 1);
						float currentPos = mBatteryLevel - batteryLevels.get(i - 1);

						float ratio = currentPos / range;

						paint.setColor(ColorUtils.blendARGB(batteryColors[i - 1], batteryColors[i], ratio));
					}
					else
					{
						paint.setColor(batteryColors[i]);
					}
					return;
				}
			}
			paint.setColor(mFGColor);
		}
		else
		{
			SweepGradient shader = new SweepGradient(centerX, centerY, mShadeColors, mShadeLevels);
			Matrix shaderMatrix = new Matrix();
			shaderMatrix.preRotate(270f, centerX, centerY);
			shader.setLocalMatrix(shaderMatrix);
			paint.setShader(shader);
		}

		mBatteryPaint.setAlpha(mAlpha);
	}

	@Override
	public void setAlpha(int alpha) {
		mAlpha = alpha;

		mFramePaint.setAlpha(Math.round(70 * alpha / 255f));

		mBoltPaint.setAlpha(alpha);
		mPowerSavePaint.setAlpha(alpha);
		mTextPaint.setAlpha(alpha);

		postInvalidate();
	}

	@SuppressLint("DiscouragedApi")
	private void updateSize()
	{
		Resources res = mContext.getResources();

		mDiameter = getBounds().bottom - getBounds().top;

		mWarningTextPaint.setTextSize(mDiameter * 0.75f);

		float strokeWidth = mDiameter / 6.5f;
		mFramePaint.setStrokeWidth(strokeWidth);
		mBatteryPaint.setStrokeWidth(strokeWidth);
		mPowerSavePaint.setStrokeWidth(strokeWidth);

		mTextPaint.setTextSize(mDiameter * 0.52f);

		mFrame.set(strokeWidth / 2.0f,
				strokeWidth / 2.0f,
				mDiameter - strokeWidth / 2.0f,
				mDiameter - strokeWidth / 2.0f);

		@SuppressLint("DiscouragedApi") 
		Path unscaledBoltPath = new Path();
		unscaledBoltPath.set(
				PathParser.createPathFromPathData(
						res.getString(
								res.getIdentifier(
										"android:string/config_batterymeterBoltPath",
										"string",
										"android"))));

		//Bolt icon
		Matrix scaleMatrix = new Matrix();
		RectF pathBounds = new RectF();

		unscaledBoltPath.computeBounds(pathBounds, true);

		float scaleF = (getBounds().height() - strokeWidth * 2) * .8f / pathBounds.height(); //scale comparing to 80% of icon's inner space

		scaleMatrix.setScale(scaleF, scaleF);

		mBoltPath = new Path();

		unscaledBoltPath.transform(scaleMatrix, mBoltPath);

		mBoltPath.computeBounds(pathBounds, true);

		//moving it to center
		mBoltPath.offset(getBounds().centerX() - pathBounds.centerX(),
				getBounds().centerY() - pathBounds.centerY());
	}

	@Override
	public void setColorFilter(@Nullable ColorFilter colorFilter) {
		mFramePaint.setColorFilter(colorFilter);
		mBatteryPaint.setColorFilter(colorFilter);
		mWarningTextPaint.setColorFilter(colorFilter);
		mBoltPaint.setColorFilter(colorFilter);
	}

	@Override
	public int getOpacity() {
		return PixelFormat.UNKNOWN;
	}

	@Override
	public int getIntrinsicHeight()
	{
		return CIRCLE_DIAMETER;
	}

	@Override
	public int getIntrinsicWidth()
	{
		return CIRCLE_DIAMETER;
	}

	@ColorInt
	private static int getColorStateListDefaultColor(Context context, int resId){
		return context.getResources().getColorStateList(resId, context.getTheme()).getDefaultColor();
	}
}

File: app/src/main/java/sh/siava/AOSPMods/Utils/batteryStyles/CircleBatteryDrawable.kt
/*
 * Copyright (C) 2017 The Android Open Source Project
 * Copyright (C) 2019 The LineageOS Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sh.siava.AOSPMods.Utils.batteryStyles

import android.content.Context
import android.content.res.Resources
import android.graphics.*
import androidx.annotation.ColorInt
import androidx.core.graphics.ColorUtils
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

class CircleBatteryDrawable(
    private val context: Context,
    frameColor: Int
) : BatteryDrawable() {

    private var fgColor: Int = Color.WHITE
    private val criticalLevel: Int
    private val warningString: String
    private val framePaint: Paint
    private val batteryPaint: Paint
    private val warningTextPaint: Paint
    private val textPaint: Paint
    private val boltPaint: Paint
    private val powerSavePaint: Paint
    private val boltPoints: FloatArray
    private val boltPath = Path()
    private val padding = Rect()
    private val frame = RectF()
    private val boltFrame = RectF()
    private val pathEffect = DashPathEffect(floatArrayOf(3f,2f),0f)

    private var iconTint = Color.WHITE
    private var intrinsicWidth: Int
    private var intrinsicHeight: Int
    private var height = 0
    private var width = 0
    private var isCharging = false
    private var powerSaveColor : Int


    private val BATTERY_STYLE_CIRCLE = 1
    private val BATTERY_STYLE_DOTTED_CIRCLE = 2

    // Dual tone implies that battery level is a clipped overlay over top of the whole shape
    private var dualTone = false

    override fun getIntrinsicHeight() = intrinsicHeight

    override fun getIntrinsicWidth() = intrinsicWidth

    private var fastCharging: Boolean = false

    override fun setFastCharging(isFastCharging: Boolean) {
        fastCharging = isFastCharging
        if(!isFastCharging)
        {
            charging = false
        }
        postInvalidate()
    }

    private var charging: Boolean = false

    override fun setCharging(isCharging : Boolean)
    {
        charging = isCharging
        if(!charging) fastCharging = false
        postInvalidate()
    }


    private var powerSaving: Boolean = false

    override fun setPowerSaveEnabled(isPowerSaving: Boolean) {
        powerSaving = isPowerSaving
        postInvalidate()
    }

    override fun refresh() {
        invalidateSelf()
    }


    private var showPercentage: Boolean = false

    override fun setShowPercent(showPercent: Boolean) {
        showPercentage = showPercent
        invalidateSelf()
    }

    private var batteryLevel: Int = -1

    override fun setBatteryLevel(mLevel: Int) {
        batteryLevel = mLevel
        postInvalidate()
    }

    private var bMeterStyle: Int = BATTERY_STYLE_CIRCLE

    override fun setMeterStyle(batteryStyle: Int) {
        bMeterStyle = batteryStyle
        postInvalidate()
    }

    // an approximation of View.postInvalidate()
    private fun postInvalidate() {
        unscheduleSelf { invalidateSelf() }
        scheduleSelf({ invalidateSelf() }, 0)
    }

    override fun setBounds(left: Int, top: Int, right: Int, bottom: Int) {
        super.setBounds(left, top, right, bottom)
        updateSize()
    }
    private var javaAlpha : Int = 255
    override fun setAlpha(alpha: Int) {
        javaAlpha = alpha
    }

    private fun updateSize() {
        val res = context.resources
        height = bounds.bottom - padding.bottom - (bounds.top + padding.top)
        width = bounds.right - padding.right - (bounds.left + padding.left)
        warningTextPaint.textSize = height * 0.75f
        intrinsicHeight = res.getDimensionPixelSize(res.getIdentifier("battery_height", "dimen", context.packageName))// res.getDimensionPixelSize(R.dimen.battery_height)
        intrinsicWidth = res.getDimensionPixelSize(res.getIdentifier("battery_height", "dimen", context.packageName))//res.getDimensionPixelSize(R.dimen.battery_height)
    }

    override fun getPadding(padding: Rect): Boolean {
        if (this.padding.left == 0 &&
                this.padding.top == 0 &&
                this.padding.right == 0 &&
                this.padding.bottom == 0
        ) {
            return super.getPadding(padding)
        }
        padding.set(this.padding)
        return true
    }

    override fun setColors(fgColor: Int, bgColor: Int, singleToneColor: Int) {
        this.fgColor = fgColor
        val fillColor = if (dualTone) fgColor else singleToneColor

        iconTint = fillColor
        framePaint.color = bgColor
//<Sia: Fixed visibility issues
//>
        invalidateSelf()
    }

    override fun draw(c: Canvas) {
        if (lastUpdate != lastVarUpdate) {
            lastUpdate = lastVarUpdate
            refreshShadeColors()
        }

        if (batteryLevel == -1) return
        val circleSize = min(width, height)
        val strokeWidth = circleSize / 6.5f
        framePaint.strokeWidth = strokeWidth
        framePaint.style = Paint.Style.STROKE
        batteryPaint.strokeWidth = strokeWidth
        batteryPaint.style = Paint.Style.STROKE
        if (bMeterStyle == BATTERY_STYLE_DOTTED_CIRCLE) {
            batteryPaint.pathEffect = pathEffect
            powerSavePaint.pathEffect = pathEffect
        } else {
            batteryPaint.pathEffect = null
            powerSavePaint.pathEffect = null
        }
        powerSavePaint.strokeWidth = strokeWidth
        frame[
                strokeWidth / 2.0f + padding.left, strokeWidth / 2.0f,
                circleSize - strokeWidth / 2.0f + padding.left
        ] = circleSize - strokeWidth / 2.0f

        setLevelPaint(batteryPaint, frame.centerX(), frame.centerY())

        if (charging || fastCharging) { // define the bolt shape

            val bl = frame.left + frame.width() / 3.0f
            val bt = frame.top + frame.height() / 3.4f
            val br = frame.right - frame.width() / 4.0f
            val bb = frame.bottom - frame.height() / 5.6f
            if (boltFrame.left != bl ||
                    boltFrame.top != bt ||
                    boltFrame.right != br ||
                    boltFrame.bottom != bb
            ) {
                boltFrame[bl, bt, br] = bb
                boltPath.reset()
                boltPath.moveTo(
                        boltFrame.left + boltPoints[0] * boltFrame.width(),
                        boltFrame.top + boltPoints[1] * boltFrame.height()
                )
                var i = 2
                while (i < boltPoints.size) {
                    boltPath.lineTo(
                            boltFrame.left + boltPoints[i] * boltFrame.width(),
                            boltFrame.top + boltPoints[i + 1] * boltFrame.height()
                    )
                    i += 2
                }
                boltPath.lineTo(
                        boltFrame.left + boltPoints[0] * boltFrame.width(),
                        boltFrame.top + boltPoints[1] * boltFrame.height()
                )
            }

            boltPaint.color = if (fastCharging && showFastCharging) fastChargingColor else fgColor
            boltPaint.alpha = javaAlpha
            c.drawPath(boltPath, boltPaint)
        }
        // draw thin gray ring first
        framePaint.alpha = (80*javaAlpha/255f).roundToInt()
        c.drawArc(frame, 270f, 360f, false, framePaint)
        // draw colored arc representing charge level
        if (batteryLevel > 0) {
            if (!charging && powerSaving) {
                powerSavePaint.alpha = javaAlpha
                c.drawArc(frame, 270f, 3.6f * batteryLevel, false, powerSavePaint)
            } else {
                batteryPaint.alpha = javaAlpha
                c.drawArc(frame, 270f, 3.6f * batteryLevel, false, batteryPaint)
            }
        }
        // compute percentage text
        if (!charging && batteryLevel != 100 && showPercentage) {
            textPaint.color = fgColor
            textPaint.textSize = height * 0.52f
            val textHeight = -textPaint.fontMetrics.ascent
            val pctText =
                    if (batteryLevel > criticalLevel)
                        batteryLevel.toString()
                    else
                        warningString
            val pctX = width * 0.5f
            val pctY = (height + textHeight) * 0.47f
            textPaint.alpha = javaAlpha
            c.drawText(pctText, pctX, pctY, textPaint)
        }
    }

    private fun setLevelPaint(paint: Paint, cx: Float, cy: Float) {
        var singleColor: Int = fgColor
        paint.shader = null
        if (fastCharging && showFastCharging && batteryLevel < 100) {
            paint.color = fastChargingColor
            return
        } else if (isCharging && showCharging && batteryLevel < 100) {
            paint.color = chargingColor
            return
        } else if (powerSaving) {
            paint.color = powerSaveColor
            return
        }

        if (!colorful || shadeColors.isEmpty()) {
            for (i in batteryLevels.indices) {
                if (batteryLevel <= batteryLevels[i]) {
                    singleColor = if (transitColors && i > 0) {
                        val range =
                            batteryLevels[i] - batteryLevels[i - 1]
                        val currentPos = batteryLevel - batteryLevels[i - 1]
                        val ratio = currentPos / range
                        ColorUtils.blendARGB(
                            batteryColors[i - 1],
                            batteryColors[i], ratio
                        )
                    } else {
                        batteryColors[i]
                    }
                    break
                }
            }
            paint.color = singleColor
        } else {
            val shader = SweepGradient(
                cx,
                cy,
                shadeColors,
                shadeLevels,
            )
            val shaderMatrix = Matrix()
            shaderMatrix.preRotate(270f, cx, cy)
            shader.setLocalMatrix(shaderMatrix)
            paint.shader = shader
            //			paint.setAlpha(128);
        }
    }

    private fun refreshShadeColors() {
        if (batteryColors == null) return
        shadeColors =
            IntArray(batteryLevels.size * 2 + 2)
        shadeLevels =
            FloatArray(shadeColors.size)
        var prev = 0f
        for (i in batteryLevels.indices) {
            val rangeLength = batteryLevels[i] - prev
            shadeLevels[2 * i] = (prev + rangeLength * .3f) / 100
            shadeColors[2 * i] = batteryColors[i]
            shadeLevels[2 * i + 1] =
                (batteryLevels[i] - rangeLength * .3f) / 100
            shadeColors[2 * i + 1] = batteryColors[i]
            prev = batteryLevels[i]
        }
        shadeLevels[shadeLevels.size - 2] =
            (batteryLevels[batteryLevels.size - 1] + (100 - batteryLevels[batteryLevels.size - 1]) * .3f) / 100
        shadeColors[shadeColors.size - 2] =
            Color.GREEN
        shadeLevels[shadeLevels.size - 1] =
            1f
        shadeColors[shadeColors.size - 1] =
            Color.GREEN
    }


    override fun setColorFilter(colorFilter: ColorFilter?) {
        framePaint.colorFilter = colorFilter
        batteryPaint.colorFilter = colorFilter
        warningTextPaint.colorFilter = colorFilter
        boltPaint.colorFilter = colorFilter
    }

    @Deprecated("Deprecated in Java",
        ReplaceWith("PixelFormat.UNKNOWN", "android.graphics.PixelFormat")
    )
    override fun getOpacity() = PixelFormat.UNKNOWN

    companion object {
        private var shadeColors: IntArray = IntArray(0)
        private var shadeLevels: FloatArray = FloatArray(0)
        private var lastUpdate: Long = -1

        private fun loadPoints(
                res: Resources,
                pointArrayRes: Int
        ): FloatArray {
            val pts = res.getIntArray(pointArrayRes)
            var maxX = 0
            var maxY = 0
            run {
                var i = 0
                while (i < pts.size) {
                    maxX = max(maxX, pts[i])
                    maxY = max(maxY, pts[i + 1])
                    i += 2
                }
            }
            val ptsF = FloatArray(pts.size)
            var i = 0
            while (i < pts.size) {
                ptsF[i] = pts[i].toFloat() / maxX
                ptsF[i + 1] = pts[i + 1].toFloat() / maxY
                i += 2
            }
            return ptsF
        }
    }

    init {
        val res = context.resources

        warningString = "!"//res.getString(R.string.battery_meter_very_low_overlay_symbol)
        criticalLevel = 5 /*res.getInteger(
                com.android.internal.R.integer.config_criticalBatteryWarningLevel
        )*/
        framePaint = Paint(Paint.ANTI_ALIAS_FLAG)
        framePaint.color = frameColor
        framePaint.isDither = true
        batteryPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        batteryPaint.isDither = true
        textPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        textPaint.typeface = Typeface.create("sans-serif-condensed", Typeface.BOLD)
        textPaint.textAlign = Paint.Align.CENTER
        warningTextPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        warningTextPaint.typeface = Typeface.create("sans-serif", Typeface.BOLD)
        warningTextPaint.textAlign = Paint.Align.CENTER


        boltPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        boltPoints =
                loadPoints(res, res.getIdentifier("batterymeter_bolt_points", "array", context.packageName))// R.array.batterymeter_bolt_points)
        powerSaveColor = getColorStateListDefaultColor(
                context,
                res.getIdentifier("batterymeter_plus_color", "color", context.packageName)//R.color.batterymeter_plus_color
        )
        powerSavePaint = Paint(Paint.ANTI_ALIAS_FLAG)
        powerSavePaint.color = powerSaveColor
        powerSavePaint.style = Paint.Style.STROKE
        intrinsicWidth = res.getDimensionPixelSize(res.getIdentifier("battery_height", "dimen", context.packageName))//R.dimen.battery_width)
        intrinsicHeight = res.getDimensionPixelSize(res.getIdentifier("battery_height", "dimen", context.packageName))//R.dimen.battery_height)

        dualTone = false//res.getBoolean(com.android.internal.R.bool.config_batterymeterDualTone)
    }

    @ColorInt
    private fun getColorStateListDefaultColor(context: Context, resId: Int): Int {
        val list = context.resources.getColorStateList(resId, context.theme)
        return list.defaultColor
    }

}